### PR TITLE
add clSetContextDestructorCallback

### LINKED ISSUE #187: Need for clSetContextReleaseCallback?
When creating a `cl_context` using either `clCreateContext` or `clCreateContextFromType`, a callback can be given. The function can optionally be given a `user_data` pointer, that will be provided to the callback when it is invoked. The callback can be called several time, and potentially after the user called `clReleaseContext`.

If the `user_data` pointer is dynamically allocated, the user has no way to know when it is safe to free this pointer. A `clSetContextReleaseCallback` could be used to solve this issue. The signature would be similar to `clSetProgramReleaseCallback`.

```c
cl_int clSetContextReleaseCallback(
    cl_context context,
    void (CL_CALLBACK* pfn_notify)(cl_context context, void* user_data),
    void* user_data);
```

### PR DESCRIPTION
**Pull request by bashbaug**
Fixes #187.

This PR adds a new API to register a callback for a context that is called when the context is destroyed.

### CODE REVIEW

File: api/opencl_platform_layer.asciidoc [hunk 1]
```diff
@@ -2205,3 +2205,52 @@ Otherwise, it returns one of the following errors:
...
+callback function on a destructor callback stack associated with _context_.
+The registered callback functions are called in the reverse order in
+which they were registered.
+If a context callback function was specified when _context_ was created,
+it will not be called after any context destructor callback is called.
+Therefore, the context destructor callback provides a mechanism for an
+application to safely re-use or free any _user_data_ specified for the
+context callback function when _context_ was created.
+
+// refError
```
Comment (kpet) [inline api/opencl_platform_layer.asciidoc:2243]: Did you intend to add that line?

Comment (bashbaug) [reply api/opencl_platform_layer.asciidoc:2243]: Confirming: this is about:

`// refError`

If so, yes, this line is included in the description of each OpenCL API.  I believe it is used by the toolchain that generates the [OpenCL reference pages](https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/).

**Comment by bashbaug** [comment]
> This also begs the (separate) question: do we want to provide an alias to `clSetProgramReleaseCallback` for consistency? I'm thinking of `clSetProgramDestructorCallback`.

Since we don't support program scope constructors and destructors in OpenCL 3.0, I think a better thing to do is to deprecate `clSetProgramReleaseCallback` and add back `clSetProgramDestructorCallback` if and only if we add functionality that requires it.

**Comment by kpet** [comment]
>Since we don't support program scope constructors and destructors in OpenCL 3.0, 

We should clarify this in the spec BTW (Appendix H is pretty clear but the normative section just says "Missing before 2.2" which implies it's supported in 3.0). I've created https://github.com/KhronosGroup/OpenCL-Docs/issues/347.

>I think a better thing to do is to deprecate clSetProgramReleaseCallback and add back clSetProgramDestructorCallback if and only if we add functionality that requires it.

Yes, I was thinking of the end state with the feature added back in, not specifically of 3.0 but I think it makes sense to deprecate in 3.0. I've created https://github.com/KhronosGroup/OpenCL-Docs/issues/348.

**Comment by keryell** [comment]
Since it is similar to some discussion we had in SYCL https://gitlab.khronos.org/sycl/Specification/-/issues/289 I wonder how this feature can be implemented safely.
For example what happens if the OpenCL runtime decide to destruct the context at exit, after `main()` exits?

**Comment by bashbaug** [comment]
Interesting discussion, thanks for passing this along.  FWIW, I don't think the SYCL proposal ([link](https://github.com/codeplaysoftware/standards-proposals/blob/master/extended-context-destruction/index.md)) is exactly the same as this feature because the proposed SYCL deleter gets called *before* the SYCL context is deleted, whereas this callback gets called *after* the OpenCL context is deleted, but there's definitely a fair amount of overlap.

> For example what happens if the OpenCL runtime decide to destruct the context at exit, after `main()` exits?

I think the simplest answer is that this behavior won't be conformant.  I'm hand-waving a little, but I imagine we would have a conformance test that ensures the context destructor callback is properly called, similar to this conformance test for `clSetMemObjectDestructorCallback`:

https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/api/test_mem_objects.cpp

Does that sound reasonable?

**Comment by bashbaug** [comment]
Merging as discussed on the July 16th teleconference.

**Comment by keryell** [comment]
Yes, there is a difference between when a `context` is deletable and when it has been actually discarded.
I am curious about the use case.
For example you create a static global C++ object that set up some OpenCL context in the constructor.
Probably it might be too late to be useful if the context destruction call-back is called on this static object destruction...

**Pull request merged**